File: metatoolkit/volcano.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import argparse
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import seaborn as sns
import os
from pathlib import Path
from adjustText import adjust_text

def volcano(df, hue=None, change='Log2FC', sig='MWW_pval', fc=1, pval=0.05,
            annot=True, ax=None, size=None, figsize=(8, 6)):

    if ax is None:
        fig, ax = plt.subplots(figsize=figsize)

    lfc = df[change]
    pvals = df[sig]
    lpvals = -np.log10(pvals)

    colors = 'black'
    if hue:
        unique_vals = df[hue].unique()
        color_palette = sns.color_palette("Reds", len(unique_vals))
        color_dict = dict(zip(unique_vals, color_palette))
        colors = df[hue].map(color_dict)

    sizes = 5
    if size:
        sizes = df[size]

    ax.scatter(lfc, lpvals, c=colors, s=sizes, alpha=1)
    ax.axvline(0, color='gray', linestyle='--')
    ax.axhline(-np.log10(pval), color='red', linestyle='-')
    ax.axhline(0, color='gray', linestyle='--')
    ax.axvline(fc, color='red', linestyle='-')
    ax.axvline(-fc, color='red', linestyle='-')

    ax.set_ylabel('-log10 p-value')
    ax.set_xlabel('log2 fold change')
    ax.set_ylim(ymin=-0.1)
    x_max = np.abs(lfc).max() * 1.1
    ax.set_xlim(xmin=-x_max, xmax=x_max)

    # Remove top and right spines
    ax.spines['top'].set_visible(False)
    ax.spines['right'].set_visible(False)

    sig_species = (lfc.abs() > fc) & (pvals < pval)
    filter_df = df[sig_species]

    if annot:
        texts = []
        for idx, row in filter_df.iterrows():
            texts.append(ax.text(row[change], -np.log10(row[sig]), s=idx, fontsize=6))
        adjust_text(texts, ax=ax, arrowprops=dict(arrowstyle='-', color='gray', lw=0.5))

    return ax


def parse_args():
    parser = argparse.ArgumentParser(description='Volcano - Produces a Volcano plot of a given dataset')
    parser.add_argument('subject', type=str, help='Path to the input TSV file')
    parser.add_argument('--change', type=str, default='coef', help='Column name for log2 fold change')
    parser.add_argument('--sig', type=str, default='qval', help='Column name for significance values (p/q)')
    parser.add_argument('--fc', type=float, default=1.0, help='Fold change threshold')
    parser.add_argument('--pval', type=float, default=0.05, help='P-value threshold')
    parser.add_argument('--annot', action=argparse.BooleanOptionalAction, default=True, help='Annotate significant points')
    parser.add_argument('--figsize', type=float, nargs=2, default=[8, 6], help='Figure size in inches (width height)')
    parser.add_argument('--output', type=str, help='Output path for the SVG file')
    return parser.parse_args()

def main():
    args = parse_args()
    df = pd.read_csv(args.subject, sep='\t', index_col=0)
    volcano(df, change=args.change, sig=args.sig, fc=args.fc, pval=args.pval,
            annot=args.annot, figsize=tuple(args.figsize))

    if args.output:
        outpath = Path(args.output)
    else:
        outname = Path(args.subject).stem if os.path.isfile(args.subject) else args.subject
        outpath = Path(f'results/{outname}_volcano.svg')

    plt.savefig(outpath)

if __name__ == '__main__':
    main()

File: metatoolkit/volcano.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(ggplot2)
  library(ggrepel)
  library(optparse)
  library(readr)
  library(dplyr)
  library(tools)
})

# Define CLI options
option_list <- list(
  make_option(c("-c", "--change"), type = "character", default = "coef", help = "Column for log2 fold change"),
  make_option(c("-s", "--sig"), type = "character", default = "qval", help = "Column for p/q-values"),
  make_option(c("--fc"), type = "double", default = 1.0, help = "Fold change threshold"),
  make_option(c("--pval"), type = "double", default = 0.05, help = "P-value threshold"),
  make_option(c("--no_annot"), action = "store_true", default = FALSE, help = "Disable annotation")
)

parser <- OptionParser(usage = "usage: %prog [options] input.tsv", option_list = option_list)
arguments <- parse_args(parser, positional_arguments = TRUE)

if (length(arguments$args) != 1) {
  stop("Exactly one input file must be provided.")
}

input_file <- arguments$args[1]
opts <- arguments$options

# Load data
df <- read_tsv(input_file, col_types = cols())
change_col <- opts$change
sig_col <- opts$sig
fc_thresh <- opts$fc
pval_thresh <- opts$pval
annotate <- !opts$no_annot

# Check required columns
if (!(change_col %in% colnames(df)) || !(sig_col %in% colnames(df))) {
  stop("Change or significance column not found in input data.")
}

# Prepare data
df <- df %>%
  mutate(log_pval = -log10(.data[[sig_col]]),
         significant = abs(.data[[change_col]]) > fc_thresh & .data[[sig_col]] < pval_thresh,
         label = ifelse(significant & annotate, rownames(df), NA))

# Plot
p <- ggplot(df, aes(x = .data[[change_col]], y = log_pval)) +
  geom_point(aes(color = significant), alpha = 0.5) +
  geom_vline(xintercept = c(-fc_thresh, 0, fc_thresh), color = c("red", "gray", "red"), linetype = c("solid", "dashed", "solid")) +
  geom_hline(yintercept = -log10(pval_thresh), color = "red", linetype = "solid") +
  scale_color_manual(values = c("black", "red")) +
  labs(x = "log2 fold change", y = "-log10 p-value") +
  theme_minimal()

if (annotate) {
  p <- p + geom_text_repel(aes(label = label), size = 2.5, max.overlaps = Inf)
}

# Save plot
outname <- paste0(file_path_sans_ext(basename(input_file)), "volcano.svg")
ggsave(outname, p, width = 8, height = 6)


